File: src/absractions/data/ISyncable_data_source.ts
import { ISyncMetadata } from "../metadata/ISync_metadata";

export interface ISyncalbeDataSource<T> {
  add(entity: T): Promise<T>;
  update(query: any, entity: T): Promise<T | null>;
  findById(id: string): Promise<T>;
  getAll(): Promise<T[]>;
  query(query: any,): Promise<T[]>;
  count(): Promise<number>;
  updateMany(entities: T[]): Promise<T[]>;
  addMany(entities: T[]): Promise<T[]>;
  fetchMany(syncMetadata: ISyncMetadata): Promise<T[]>;
  dispose(): void;
}

File: fast_sync_dart_client/fast_sync_core/lib/src/absraction/data/isyncable_data_source.dart
import 'package:fast_sync_client/src/absraction/metadata/isync_metadata.dart';

abstract class ISyncableDataSource<T> {
  Future<T> add(T entity);
  Future<T> update(dynamic query, T entity);
  Future<T> findById(String id);
  Future<List<T>> getAll();
  Future<List<T>> query(dynamic query);
  Future<int> count();
  Future<List<T>> updateMany(List<T> entities);
  Future<List<T>> addMany(List<T> entities);
  Future<List<T>> fetchMany(ISyncMetadata syncMetadata);
  void dispose();
}
